wal/wal.go

```diff
@@ -299,17 +299,55 @@ func OpenForRead(lg *zap.Logger, dirpath string, snap walpb.Snapshot) (*WAL, err
 }
 
 func openAtIndex(lg *zap.Logger, dirpath string, snap walpb.Snapshot, write bool) (*WAL, error) {
```

Comment:
why do we need to split the openAtIndex func?

Reply:
The way the decoder works is that once it traverses through the WAL records, it depletes the WAL file readers, such that the same decoder cannot be used again.

https://github.com/etcd-io/etcd/blob/6da17cda18307eff6e946c824fb7b64ecaf34fbe/wal/decoder.go#L70

The `Verify` method needs to traverse through the WAL files in such a way that it does not affect the subsequent `ReadAll` call (which uses the decoder). This is why the `Verify` method re-creates a decoder from the WAL files and uses that. And to achieve this, `openAtIndex` function had to be split so that the code for creating the decoder can be re-used by `Verify` method.